src/Opaleye/TH.hs

```diff
@@ -0,0 +1,652 @@
+        updateExp <- [e|liftIO $ Prelude.head <$> runUpdateReturning conn $(return $ VarE $ mkName $ makeTablename t) (\_ -> toWrite $ constant model) (\r -> ($(return $ VarE $ mkName pField) r) .== (constant $  $(return $ VarE $ mkName pField) model)) id |]
+        deleteExp <- [e|liftIO $ runDelete conn $(return $ VarE $ mkName $ makeTablename t) (\r -> ($(return $ VarE $ mkName pField) r) .== (constant $  $(return $ VarE $ mkName pField) model)) |]
+        let pat = [VarP $ mkName "conn", VarP $ mkName "model"]
+        let insertFunc = FunD (mkName "insertModel") [Clause pat (NormalB insertExp) convertToPgWrite]
+        let updateFunc = FunD (mkName "updateModel") [Clause pat (NormalB updateExp) convertToPgWrite]
+        let deleteFunc = FunD (mkName "deleteModel") [Clause pat (NormalB deleteExp) []]
+        return [InstanceD Nothing [] (AppT (ConT $ mkName "DbModel") (ConT $ mkName r)) [insertFunc, updateFunc, deleteFunc]]
+      where
+        makeConversionFunction :: Q [Dec]
+        makeConversionFunction = do
+          let
+            pgReadType = ConT $ (mkName $ makePGReadTypeName r)
+            pgWriteType = ConT $ (mkName $ makePGWriteTypeName r)
+            conversionFunctionSig = SigD (mkName "toWrite") $ AppT (AppT ArrowT pgReadType) pgWriteType
+            conversionFunction = FunD (mkName "toWrite") [Clause [makePattern] (NormalB conExp) []]
+          return $ [conversionFunctionSig, conversionFunction]
+          where
+            conExp :: Exp
+            conExp = foldl AppE (ConE $ mkName $ r) $ getFieldExps
+            getFieldExps :: [Exp]
+            getFieldExps = zipWith getFieldExp fieldInfos [1..]
+              where
+              getFieldExp :: ColumnInfo -> Int -> Exp
+              getFieldExp ci ix = if (columnDefault ci) then (AppE (ConE $ mkName "Just") mkVarExp) else mkVarExp
+                where
+                  mkVarExp :: Exp
+                  mkVarExp = VarE $ mkName $ "a" ++ (show ix)
+            makePattern :: Pat
+            makePattern = ConP (mkName $ r) $ VarP <$> (mkName.(\x -> ('a':x)).show) <$>  take (Prelude.length fieldInfos) [1..]
+        getPrimaryKeyField :: String -> String -> (Maybe String)
+        getPrimaryKeyField tname modelName = case (filter (\ci -> columnPrimary ci)) fieldInfos of
+            [primaryField] -> Just $ makeFieldName modelName (columnName primaryField)
+            _ -> Nothing
+  
+makeTablename :: String -> String
+makeTablename t = t ++ "Table"
+
+makePGReadTypeName :: String -> String
+makePGReadTypeName tn = tn ++ "PGRead"
+
+makePGWriteTypeName :: String -> String
+makePGWriteTypeName tn = tn ++ "PGWrite"
+
+makeAdapterName :: String -> String
+makeAdapterName tn = 'p':tn
+
+makeOpaleyeModels :: (ConnectInfo, Options) -> Q [Dec]
+makeOpaleyeModels (ci, op) = makeOpaleyeModels' (ci, op, [])
+
+makeOpaleyeModels' :: Env -> Q [Dec]
+makeOpaleyeModels' env = fst <$> runStateT (do
```

Comment:
Is there a way to write this without the `StateT` monad? Perhaps collect all the "operations" in a large ADT first, then filter out the duplicates, then convert the ADT to a TH splice?